### DISCUSSION TITLE
Real example of a project structure in Evennia?

### DISCUSSION THREAD

**Discussion by RodRitter** [evennia/evennia#2691] [Evennia Questions]
Hey all,

I've been messing around with Evennia for a while but really want to start planning a scalable project and I'm trying to find a good way to structure the project. Are there any open source projects that I could take a look at to get an idea of a scalable project structure?

My inclination is to create folders for custom `typeclasses` like this. Is it a feasible pattern?
```
- [typeclasses]
    - rooms.py 
    - characters.py
    - ...
    - [game_rooms]
        - chargen.py
        - temple.py
        - ...
    - [game_characters]
        - character.py
        - npc.py
        - god.py
        - ...
```

Another thing I'd like to know is how screens, tables and ascii art is organized. At the moment I'm adding everything in the command func itself, but it doesn't seem scalable. Perhaps a template system?
```
"""
    templates.py
"""
def score(name, level):
    table = evtable.EvTable(f"You are: {name}", f"Level {level}"
```
```
"""
    player_cmds.py
"""
class CmdScore(Command):
    key = "score"
    aliases = ["sc"]
    lock = "cmd:all()"
    help_category = "General"

   def func(self):
        score_table = templates.score(self.caller.name, self.caller.db.level)
        self.caller.msg(sc_table)
```
I'd love to see how others structure their projects

**Comment by Tegiminis** [comment]
I must admit, my folder structure is totally fucked right now, as I have a lot of non-typeclassed modules in the typeclasses folder, hah. But I plan on refactoring it to this format:

- All "general" typeclasses are stored in the typeclasses folder (character, room, etc). General typeclasses of the same type are stored in a module together (for example: NPC and Player are both stored in `character.py`)
- Anything that is assigned to a typeclass but isn't a typeclass is stored in a components folder within typeclasses (for example, `buff.py` which contains the buff and buffhandler definitions)
- "Rules" and other elements which are typeclass-agnostic are stored in the "world" folder (for example: loot rolling is done through `loot.py`). This is what the development tutorial recommends I believe.
- Specific content which cannot be done through prototypes is bundled together in a "content" folder and grouped into similar modules (for example, Weakness and Blind debuffs are both in `debuffs.py`)

So the structure might look something like:

```
typeclasses
    components
        buff.py
        cooldown.py
        ...
    character.py
    room.py
    item.py
    weapon.py
    ...
world
    rules.py
    loot.py
    chargen.py
    ...
content
    debuffs.py
    npc_weapons.py
    ...
```

---

As for tables, I'm afraid I can't help you there. At the moment I don't do any formatting, I just send single lines out. I know other people definitely have some advice on that front. The way you set it up in your example seems perfectly fine to me though.
